python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py

```diff
@@ -0,0 +1,184 @@
+from typing import Callable, Optional, List, Dict, TYPE_CHECKING
+
+import ray
+from ray.data._internal.remote_fn import cached_remote_fn
+from ray.data._internal.memory_tracing import trace_allocation
+from ray.data._internal.execution.interfaces import (
+    RefBundle,
+)
+from ray.data._internal.execution.util import merge_ref_bundles
+from ray.data.block import Block, BlockAccessor, BlockExecStats
+from ray.data.context import DatasetContext
+from ray.types import ObjectRef
+from ray._raylet import ObjectRefGenerator
+
+if TYPE_CHECKING:
+    from ray.data._internal.execution.operators.map_operator import MapOperator
+
+
+def _map_task(fn: Callable, *blocks: List[Block]):
+    """Remote function for a single operator task.
+
+    Args:
+        fn: The callable that takes Iterator[Block] as input and returns
+            Iterator[Block] as output.
+        blocks: The concrete block values from the task ref bundle.
+
+    Returns:
+        A generator of blocks, followed by the list of BlockMetadata for the blocks
+        as the last generator return.
+    """
+    output_metadata = []
+    stats = BlockExecStats.builder()
+    for b_out in fn(blocks):
+        m_out = BlockAccessor.for_block(b_out).get_metadata([], None)
+        m_out.exec_stats = stats.build()
+        output_metadata.append(m_out)
+        yield b_out
+        stats = BlockExecStats.builder()
+    yield output_metadata
+
+
+class _TaskState:
+    """Tracks the driver-side state for an MapOperator task.
+
+    Attributes:
+        inputs: The input ref bundle.
+        output: The output ref bundle that is set when the task completes.
+    """
+
+    def __init__(self, inputs: RefBundle):
+        self.inputs: RefBundle = inputs
+        self.output: Optional[RefBundle] = None
+
+
+class MapOperatorTasksImpl:
+    def __init__(self, op: "MapOperator"):
+        # Execution arguments.
+        self._op = op
+        self._transform_fn = op.get_transform_fn()
+        self._ray_remote_args = op.ray_remote_args()
+
+        # The temporary block bundle used to accumulate inputs until they meet the
+        # min_rows_per_batch requirement.
+        self._block_bundle: Optional[RefBundle] = None
+        self._min_rows_per_batch: int = op._min_rows_per_batch
+
+        # Execution state.
+        self._tasks: Dict[ObjectRef[ObjectRefGenerator], _TaskState] = {}
+        self._tasks_by_output_order: Dict[int, _TaskState] = {}
+        self._input_deps_done: bool = 0
+        self._next_task_index: int = 0
+        self._next_output_index: int = 0
+        self._obj_store_mem_alloc: int = 0
+        self._obj_store_mem_freed: int = 0
+        self._obj_store_mem_cur: int = 0
+        self._obj_store_mem_peak: int = 0
+
+    def add_input(self, bundle: RefBundle) -> None:
+        if self._min_rows_per_batch is None:
+            self._create_task(bundle)
+            return
```

Comment:
Couldn't we eliminate this short-circuit by setting `self._min_rows_per_batch: int = op._min_rows_per_batch or 0` in the constructor? `get_num_rows` is not expensive, and we should always fall through to creating a task on the current bundle. With the suggestion that I have below, we'll then only have a single `self._create_task()` call in `add_input()`.